frontend/lib/resolution/prims.cpp

```diff
@@ -511,6 +526,255 @@ static QualifiedType primObjectToInt(Context* context, const CallInfo& ci) {
+  }
+  return makeParamBool(context, eval);
+}
+
+static QualifiedType
+primIsAbsEnumType(Context* context, const CallInfo& ci) {
+  if (ci.numActuals() < 1) return {};
+  bool eval = false;
+  if (auto t = ci.actual(0).type().type()) {
+    if (auto et = t->toEnumType()) eval = et->isAbstract();
+  }
+  return makeParamBool(context, eval);
+}
+
+static QualifiedType
+primIsPod(Context* context, const CallInfo& ci) {
+  CHPL_UNIMPL("PRIM_IS_POD");
+  return {};
+}
+
+static QualifiedType
+primIsCoercible(Context* context, const CallInfo& ci) {
+  if (ci.numActuals() < 2) return {};
+
+  // In the old compiler this primitive only considers 'type'.
+  const auto tpFrom = ci.actual(0).type().type();
+  const auto tpTo = ci.actual(1).type().type();
+  const auto qtFrom = QualifiedType(QualifiedType::TYPE, tpFrom);
+  const auto qtTo = QualifiedType(QualifiedType::TYPE, tpTo);
+
+  const auto canPass = CanPassResult::canPass(context, qtFrom, qtTo);
+  const bool eval = canPass.passes() &&
+                    (canPass.instantiates() || canPass.converts()) &&
+                    !canPass.promotes();
+  return makeParamBool(context, eval);
+}
+
+static std::string typeToString(Context* context, const Type* t);
+
+static void
+doRegularCompositeTypeToString(Context* context, const CompositeType* ct,
+                               const ClassType* asClassType,
+                               std::stringstream& ss) {
+  bool shouldAddNilableQuestionMark = false;
+
+  // Handle class management and check for nilability.
+  if (auto cls = asClassType) {
+    auto d = cls->decorator();
+    shouldAddNilableQuestionMark = d.isNilable();
+    if (d.isBorrowed()) {
+      ss << "borrowed ";
+    } else if (d.isUnmanaged()) {
+      ss << "unmanaged ";
+    } else if (d.isManaged()) {
+      auto m = cls->manager();
+      CHPL_ASSERT(m);
+      // Just print the manager type name (they're all builtins for now).
+      ss << typeToString(context, cls->manager()) << " ";
+    }
+  }
+
+  ss << ct->name().c_str();
+
+  if (shouldAddNilableQuestionMark) ss << "?";
+
+  // Recursively print substitutions.
+  if (!ct->substitutions().empty()) {
+    auto sortedSubs = ct->sortedSubstitutions();
+    ss << "(";
+    for (size_t i = 0; i < sortedSubs.size(); i++) {
+      auto& qt = sortedSubs[i].second;
+      ss << typeToString(context, qt.type());
+      const bool last = (i+1) == sortedSubs.size();
+      if (!last) ss << ", ";
+    }
+    ss << ")";
+  }
+}
+
+static std::string typeToString(Context* context, const Type* t) {
+  std::stringstream ss;
+
+  // TODO: Should stringify be able to handle everything?
+  const bool canSimpleStringify = t->isPrimitiveType() ||
+                                  t->isBuiltinType() ||
+                                  t->isStringType() ||
+                                  t->isBytesType() ||
+                                  t->isEnumType() ||
+                                  t->isAnyType();
```

Comment:
The answer to your todo is, IMO, "yes" -- you're going through a lot of trouble to imitate here the logic that `stringify(CHPL_SYNTAX)` should already be going through. 

I thought at first the problem was the `context` argument, but upon inspection, it seems to only be used in the mutual recursion between `typeToString` and `doRegularCompositeTypeToString`, and nowhere else. So your logic doesn't need a context.

I really think you should change `stringify` to do the right thing under `CHPL_SYNTAX`.

Reply:
Yes, I agree. I'll leave that as followup.